pkg/controller/release/strategy_executor.go

```diff
@@ -2,44 +2,37 @@ package release
 
 import (
 	"fmt"
-	"sort"
 	"time"
 
-	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/equality"
-	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/client-go/tools/record"
 	"k8s.io/klog"
 
 	shipper "github.com/bookingcom/shipper/pkg/apis/shipper/v1alpha1"
 	"github.com/bookingcom/shipper/pkg/controller"
 	shippererrors "github.com/bookingcom/shipper/pkg/errors"
 	"github.com/bookingcom/shipper/pkg/util/conditions"
-	diffutil "github.com/bookingcom/shipper/pkg/util/diff"
-	releaseutil "github.com/bookingcom/shipper/pkg/util/release"
 )
 
 type StrategyExecutor struct {
 	curr, prev, succ *releaseInfo
-	recorder         record.EventRecorder
 	hasIncumbent     bool
 }
 
-func NewStrategyExecutor(curr, prev, succ *releaseInfo, recorder record.EventRecorder, hasIncumbent bool) *StrategyExecutor {
+func NewStrategyExecutor(curr, prev, succ *releaseInfo, hasIncumbent bool) *StrategyExecutor {
 	return &StrategyExecutor{
 		curr:         curr,
 		prev:         prev,
 		succ:         succ,
-		recorder:     recorder,
 		hasIncumbent: hasIncumbent,
 	}
 }
 
-type PipelineContinuation bool
+type PipelineContinuation int
 
 const (
-	PipelineBreak    PipelineContinuation = false
-	PipelineContinue                      = true
+	PipelineBreak PipelineContinuation = iota
+	PipelineContinue
+	PipelineComplete
```

Comment:
I'm a bit concerned about this approach as it might introduce some unwanted coupling between pipeline stages, i.e. steps will know about each other. The initial idea of a regular pipeline step was: can you proceed this state? yes: return pipeline continue; no: return break as it makes no sense to do anything with it. Maybe I over-emphasised the actual meaning of Break, which simply meant to break the loop but didn't mean a negative meaning exclusively.

Reply:
@icanhazbroccoli I reverted these changes, and instead just made sure that `genReleaseStrategyStateEnforcer` returns `PipelineContinue`. If the last step in a pipeline returns that, it is meant to be success, so that addresses the implicit dependency between stages.